agent/src/plugin_manager/device_plugin_instance_controller.rs

```diff
@@ -0,0 +1,1517 @@
+            device_manager,
+        }
+    }
+
+    pub async fn free_slot(&self, device_id: String) -> Result<(), DevicePluginError> {
+        let (plugin_name, slot_id) = device_id
+            .rsplit_once('-')
+            .ok_or(DevicePluginError::UsageParseError)?;
+        let slot_id = slot_id
+            .parse::<usize>()
+            .map_err(|_| DevicePluginError::UsageParseError)?;
+        {
+            let plugin = self.instance_plugins.lock().await.get(plugin_name).cloned();
+            if let Some(plugin) = plugin {
+                return plugin.free_slot(slot_id).await;
+            }
+        }
+        {
+            let plugin = self
+                .configuration_plugins
+                .lock()
+                .await
+                .get(plugin_name)
+                .cloned();
+            if let Some(plugin) = plugin {
+                return plugin.free_slot(slot_id).await;
+            }
+        }
+        Err(DevicePluginError::NoSlot)
+    }
+
+    pub async fn get_used_slots(&self) -> HashSet<String> {
+        let mut slots: HashSet<String> = Default::default();
+        for (instance, plugin) in self.instance_plugins.lock().await.iter() {
+            slots.extend(
+                plugin
+                    .slots_status
+                    .lock()
+                    .await
+                    .borrow()
+                    .iter()
+                    .enumerate()
+                    .filter_map(|(i, u)| match u {
+                        DeviceUsage::Node(n) if *n == self.node_name => {
+                            Some(format!("akri.sh/{}-{}", instance, i))
+                        }
+                        DeviceUsage::Configuration { vdev, node } if *node == self.node_name => {
+                            Some(vdev.to_string())
+                        }
+                        _ => None,
+                    }),
+            );
+        }
+        slots
+    }
+}
+
+pub fn start_dpm(dpm: Arc<DevicePluginManager>) -> (Store<Instance>, JoinHandle<()>) {
+    let api = dpm.kube_client.all().as_inner();
+    let controller = Controller::new(api, Default::default());
+    let store = controller.store();
+    let task = tokio::spawn(async {
+        controller
+            .graceful_shutdown_on(async {
+                let mut signal =
+                    tokio::signal::unix::signal(tokio::signal::unix::SignalKind::terminate())
+                        .unwrap();
+                signal.recv().await;
+            })
+            .run(reconcile, error_policy, dpm)
+            .for_each(|_| futures::future::ready(()))
+            .await
+    });
+    (store, task)
+}
+
+pub async fn reconcile(
+    instance: Arc<Instance>,
+    ctx: Arc<DevicePluginManager>,
+) -> Result<Action, DevicePluginError> {
+    trace!("Plugin Manager: Reconciling {}", instance.name_any());
+    let api = ctx.kube_client.namespaced(&instance.namespace().unwrap());
+    if !instance.spec.nodes.contains(&ctx.node_name)
```

Comment:
When you have a shared Instance, if one of the node DH is no longer able to discover the device (say partial network issue), it will get removed from the list without the whole Instance getting deleted